airbyte-integrations/connectors/source-freshdesk/source_freshdesk/api.py

```diff
@@ -138,15 +138,58 @@ def _api_get(self, url: str, params: Mapping = None):
     def list(self, fields: Sequence[str] = None) -> Iterator[dict]:
         """Iterate over entities"""
 
+    @property
+    def name(self):
+        """Name of the stream"""
+        stream_name = self.__class__.__name__
+        if stream_name.endswith("API"):
+            stream_name = stream_name[:-3]
+        return stream_name
+
     def read(self, getter: Callable, params: Mapping[str, Any] = None) -> Iterator:
         """Read using getter"""
         params = params or {}
-        for page in range(1, self.maximum_page):
-            batch = list(getter(params={**params, "per_page": self.result_return_limit, "page": page}))
-            yield from batch
 
-            if len(batch) < self.result_return_limit:
-                break
+        stream = self.name
+        # This block extends TicketsAPI Stream to overcome '300 page' server error.
+
+        if stream == "Tickets":
+            page_break = 300  # the maximum page allowed to pull during pagination.
+
+            for page in range(1, self.maximum_page):
```

Comment:
what about something like this instead?

```
        
        if stream == "Tickets":
            ticket_paginate_limit = 300  # the maximum page allowed to pull during pagination.

            for page in range(1, ticket_paginate_limit):
                batch = list(getter(params={**params, "per_page": self.result_return_limit, "page": page}))
                yield from batch
                last_record = batch[0]["updated_at"]  # position 0, because the records are returned in 'desc'

                if len(batch) < self.result_return_limit:
                    return iter()
            
            for more_page in range(1, self.maximum_page):
                last_record = pendulum.parse(last_record).add(seconds=1)
                batch = list(getter(
                                    params={
                                        **params,
                                        "order_by": "updated_at",
                                        "updated_since": last_record,
                                        "per_page": self.result_return_limit,
                                        "page": more_page,
                                    }
                                )
                            )
                yield from batch
                
                if len(batch) < self.result_return_limit:
                    return iter()

        for page in range(1, self.maximum_page):
            batch = list(getter(params={**params, "per_page": self.result_return_limit, "page": page}))
            yield from batch

            if len(batch) < self.result_return_limit:
                return iter()
```
* instead of breaking, return an empty Iterator.
* avoid the nested if/else and loops
* much easier to follow what is happening for the tickets stream

we can even pull out the ticket stream read into another function to keep the read function clean.

I'm not a Python expert so this is just a suggestion.

Reply:
`instead of breaking, return an empty Iterator.` 
Honestly, while working on this, forgot we can use this, but an empty iterator should accept at least 1 arg, so should be like this: `return iter( () )`

`avoid the nested if/else and loops`
Indeed, much more readable. Thanks.

`we can even pull out the ticket stream read into another function to keep the read function clean.`
Probably yes, however, if we modify the existing class of TicketsAPI() and incorporate this code into the class, we are passing usage of Incremental options like stream_state. I've tried both ways. 
The way it represented now in PR - doesn't break anything, only extends the functionality of the TicketsAPI stream on the way of standard pagination that is used by Incremental as well. Probably there is a better way of how to fix this, this was the quick fix. 

I've tested it like this:
- changed the `result_return_limit = 3`, means we return only 3 records/page
- placed the `print` statements in the fix-code-block just to check the intermediate outputs between the streams:
   q: what stream, what page, what state, what last_record we fetched, on what page we switch the flow of pagination, 
       etc? 
- changed the `ticket_paginate_limit = 2`, which means we switch the flow of pagination (using new parameters) from the next page.
- changed `last_record = batch[0]["updated_at"]`
       * we would receive 2 pages in total, page 1 with 3 results, page 2 would be with 0 results (because we have 3 
       tickets output) 
- changed `last_record = batch[1]["updated_at"]`
       * we would receive 2 pages in total, page 1 with 3 records, page 2 would be with 1 records (because we have 
       3 tickets output)
- changed `last_record = batch[-1]["updated_at"]`
       * we would receive 2 pages in total, page 1 with 3 records, page 2 would be with 2 records (because we have 
       3 tickets output)

I didn't create some extra unit_tests for this, because in order to catch this error from the server, we should have more than 300 pages of tickets in the sandbox, which is not necessary, i believe, because this pagination fix should overcome the bug, by switching the state of parameters after `ticket_paginate_limit = 300`.

Reply:
Makes sense. Not splitting the function is fine.

I agree inserting 300 pages of dummy data is not a good use of time. What if we used mocks? I believe you when you said this is manually tested. The reason this makes me nervous is that someone else modifying this code later on might not know/be able to do the same as you did and introduce a regression bug.